scripts/gha/test_lab.py

```diff
@@ -0,0 +1,334 @@
+    " to get this id. If none, will use FTL's default.")
+flags.DEFINE_string(
+    "android_api", None,
+    "API level for desired device. See module docstring for details on how"
+    " to find available values. If none, will use FTL's default.")
+flags.DEFINE_string(
+    "ios_model", None,
+    "Model id for desired device. See module docstring for details on how"
+    " to get this id. If none, will use FTL's default.")
+flags.DEFINE_string(
+    "ios_version", None,
+    "iOS version for desired device. See module docstring for details on how"
+    " to find available values. If none, will use FTL's default.")
+
+
+def main(argv):
+  if len(argv) > 1:
+    raise app.UsageError("Too many command-line arguments.")
+
+  testapp_dir = FLAGS.testapp_dir
+  code_platform = FLAGS.code_platform
+
+  android_device = Device(model=FLAGS.android_model, version=FLAGS.android_api)
+  ios_device = Device(model=FLAGS.ios_model, version=FLAGS.ios_version)
+
+  _authorize_gcs(FLAGS.key_file)
+
+  testapps = []
+  for file_dir, _, file_names in os.walk(testapp_dir):
+    for file_name in file_names:
+      full_path = os.path.join(file_dir, file_name)
+      if file_name.endswith(".apk"):
+        testapps.append((android_device, _ANDROID, full_path))
+      elif file_name.endswith(".ipa"):
+        testapps.append((ios_device, _IOS, full_path))
+  logging.info("Testapps found: %s", "\n".join(path for _, _, path in testapps))
+
+  gcs_base_dir = _get_base_results_dir()
+  logging.info("Storing results in %s", _relative_path_to_gs_uri(gcs_base_dir))
+
+  tests = []
+  for device, platform, path in testapps:
+    # e.g. /testapps/unity/firebase_auth/app.apk -> unity_firebase_auth_app_apk
+    name = path[len(testapp_dir) + 1:].replace("/", "_").replace(".", "_")
+    tests.append(
+        Test(
+            device=device,
+            platform=platform,
+            testapp_path=path,
+            results_dir=gcs_base_dir + "/" + name))
+
+  logging.info("Sending testapps to FTL")
+  threads = []
+  for test in tests:
+    thread = threading.Thread(target=test.run)
+    threads.append(thread)
+    thread.start()
+  for thread in threads:
+    thread.join()
+
+  # Useful diagnostic information to debug unexpected errors involving things
+  # not being where they're supposed to be. This will show everything generated
+  # in the GCS bucket for this run. Includes video links, binaries, and results.
+  logging.info(
+      "Results directory:\n%s",
+      "\n".join(_gcs_list_dir(_relative_path_to_gs_uri(gcs_base_dir)))
+  )
+
+  all_success = _report_results(tests, code_platform)
+  return 0 if all_success else 1
+
+
+def _get_base_results_dir():
+  """Defines the object used on GCS for all tests in this run."""
+  # We generate a unique directory to store the results by appending 4
+  # random letters to a timestamp. Using a timestamp is useful so that the
+  # the directory for different runs get sorted base on when it was run.
```

Comment:
nit: gets sorted based

Reply:
Also two consecutive "the"s. Three mistakes in one sentence, must have been one of those days.